### PR TITLE
fix(moe): Avoid hang in TEGroupedMLP when a rank has zero local tokens

### PR DESCRIPTION
**Pull request by jubick1337**
When a rank receives zero locally-routed tokens (a common edge case under heavy EP imbalance or small global batches), TEGroupedMLP.forward used to skip linear_fc1 / linear_fc2 entirely while peer ranks still executed them. The resulting autograd graphs diverged across ranks, causing the next DDP / EP collective to hang. Some GroupedGEMM backends also fail outright on empty input.

Detect the zero-local-tokens condition right after tokens_per_expert is materialized as a Python list. Synthesize one zero-vector token per local expert with a zero routing probability, run the regular forward path, and then slice the synthetic rows back off before returning so the caller's empty-output contract is preserved. Slicing keeps the autograd hooks on the linear modules attached so backward fires on this rank just like on peers, which is what re-aligns the cross-rank graphs.

The fused-impl early return path is left untouched.

# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/transformer/moe/experts.py [hunk 1]
```diff
@@ -558,6 +558,27 @@ def forward(
...
+        if is_dummy_forward:
+            dummy_hidden = permuted_local_hidden_states.new_zeros(
+                (self.num_local_experts, permuted_local_hidden_states.shape[1])
+            )
+            dummy_probs = permuted_probs.new_zeros((self.num_local_experts,))
+            permuted_local_hidden_states = torch.cat(
+                [permuted_local_hidden_states, dummy_hidden], dim=0
+            )
+            permuted_probs = torch.cat([permuted_probs, dummy_probs], dim=0)
+            tokens_per_expert = [1] * self.num_local_experts
```
Comment (Victarry) [inline megatron/core/transformer/moe/experts.py:580]: Hi, @jubick1337, thanks for the PR! 
But this fix is kind of hack, actually it's better to put the fix in TE to make sure every operation will work and have backward gradient even the number of tokens is zero. We fixed similar issues in this way before as in https://github.com/NVIDIA/TransformerEngine/pull/648

Comment (jubick1337) [reply megatron/core/transformer/moe/experts.py:580]: Thanks for your feedback @Victarry !
Opening [PR](https://github.com/NVIDIA/TransformerEngine/pull/3019) to TE
